src/rootPages/Designer/ui_work_interface_list_copyPage.js

```diff
@@ -0,0 +1,453 @@
+               .pages()
+               .forEach(function (p) {
+                  addPage(p, indent + "-");
+               });
+         };
+         // this.currentApplication.pages((p) => p instanceof AB.Class.ABViewPage).forEach(
+         this.currentApplication.pages().forEach(function (page) {
+            addPage(page, "");
+         });
+
+         $$(this.ids?.parentList).define("options", options);
+         $$(this.ids?.parentList).refresh();
+      }
+
+      /**
+       * @function hide()
+       *
+       * remove the busy indicator from the form.
+       */
+      hide() {
+         if (this.$component) this.$component.hide();
+      }
+
+      /**
+       * @function cancel()
+       *
+       * remove the form.
+       */
+      cancel() {
+         this.formClear();
+         this.emit("cancel");
+      }
+
+      /**
+       * Show the busy indicator
+       */
+      busy() {
+         if (this.$component) {
+            this.$component.showProgress();
+         }
+      }
+
+      /**
+       * Hide the busy indicator
+       */
+      ready() {
+         if (this.$component) {
+            this.$component.hideProgress();
+         }
+      }
+
+      /**
+       * @method done()
+       * Finished saving, so hide the popup and clean up.
+       * @param {interface} obj
+       */
+      done(obj) {
+         this.ready();
+         this.hide(); // hide our popup
+         this.emit("save", obj); // tell parent component we're done
+      }
+
+      /**
+       * @method save
+       * verify the current info is ok, package it, and return it to be
+       * added to the application.createModel() method.
+       * then take the data gathered, and
+       * add it to our current application.
+       * @param {obj} values  key=>value hash of model values.
+       * @return {Promise}
+       */
+      async save() {
+         var saveButton = $$(this.ids.buttonSave);
+         saveButton.disable();
+         // show progress
+         this.busy();
+
+         var Form = this.$form;
+
+         Form.clearValidation();
+
+         // if it doesn't pass the basic form validation, return:
+         if (!Form.validate()) {
+            saveButton.enable();
+            this.ready();
+            return false;
+         }
+
+         var values = Form.getValues();
+
+         // must have an application set.
+         if (!this.currentApplication) {
+            webix.alert({
+               title: L("Shoot!"),
+               test: L("No Application Set!  Why?"),
+            });
+            this.emit("save.error", true);
+            return false;
+         }
+
+         if (!values) {
+            // SaveButton.enable();
+            // CurrentEditor.formReady();
+            return;
+         }
+
+         if (values.parent === "-") {
+            values.parent = null;
+         } else if (values.parent) {
+            values.parent = this.currentApplication.urlResolve(values.parent);
```

Comment:
what is the value of `values.parent` here?  I didn't think it was a url to resolve.  Is it the actual ABApplication?

Reply:
Added a comment to clarify, but after refactoring the copy function works _worse_ ... I think this will need a review later on. not sure I can catch all the issues without the ability to get into what is going on with the widgets